### PR TITLE
Bug 1331399 - Reduce payload size and memory footprint of job data

### PR DESCRIPTION
**Pull request by camd**
This fixes the errors in the first attempt and adds unit tests to catch this in the future by ensuring (at 3 different levels) that the jobs have the fields required to do retriggers and add-new-jobs.

The first commit is the original, the second has the fixes and new tests.

### CODE REVIEW

File: ui/helpers/job.js [hunk 1]
```diff
@@ -177,7 +177,7 @@ export const findGroupInstance = function findGroupInstance(job) {
 // Fetch the React instance based on the jobId, and if scrollTo
 // is true, then scroll it into view.
 export const findJobInstance = function findJobInstance(jobId, scrollTo) {
-  const jobEl = $('.th-view-content')
+  const jobEl = $('#push-list')
```
Comment (camd) [inline ui/helpers/job.js]: Changed this to make it easier in the unit tests. It creates the same constraint to find the right elements.

File: ui/helpers/job.js [hunk 2]
```diff
@@ -237,15 +237,6 @@ export const addAggregateFields = function addAggregateFields(job) {
   return job;
 };
 
-// Remove any fields we don't need for display in the greater job field after
```
Comment (camd) [inline ui/helpers/job.js]: This was the troublesome part that caused retriggers and add-new-jobs to fail.  We need the ``signature`` and ``job_type_name`` fields on jobs.

File: ui/helpers/job.js (hunk 1 above)
Comment (sarah-clements) [reply ui/helpers/job.js]: Can you remove the jquery in this pr or are you planning to remove it all it once in another one? :)

File: treeherder/webapp/api/jobs.py [hunk 3]
```diff
@@ -96,9 +96,73 @@ class Meta:
...
+    ]
+    queryset = Job.objects.all().order_by('id').select_related(
+        *_default_select_related
+    ).values(*_query_field_names)
+    serializer_class = serializers.JobSerializer
+    filterset_class = JobFilter
+    pagination_class = pagination.JobPagination
+
+    def get_serializer_context(self):
+        option_collection_map = OptionCollection.objects.get_option_collection_map()
```
Comment (sarah-clements) [inline treeherder/webapp/api/jobs.py]: Is this getting called for every iteration over the queryset, or is it called once and then cached?

File: ui/models/job.js [hunk 4]
```diff
@@ -20,22 +20,20 @@ export default class JobModel {
...
       let nextPagesJobs = [];
 
-      // if the number of elements returned equals the page size,
-      // fetch the next pages
-      if (fetchAll && results.length === meta.count) {
-        const count = parseInt(meta.count, 10);
-        const offset = parseInt(meta.offset, 10) + count;
-        const newOptions = { ...options, offset, count };
+      // if ``next`` gives a URL for more, fetch the next pages
+      if (fetchAll && next) {
```
Comment (sarah-clements) [inline ui/models/job.js]: `nextUrl` would be clearer and then maybe eliminate the need for a comment?

**Comment by armenzg** [comment]
I'm not familiar with TH's frontend to give any value in such large PR.

File: ui/models/job.js (hunk 4 above)
Comment (camd) [reply ui/models/job.js]: Good point.  Done...

File: treeherder/webapp/api/jobs.py (hunk 3 above)
Comment (camd) [reply treeherder/webapp/api/jobs.py]: It's called once every time the ``/api/job/`` endpoint is called.  The result list of options *rarely* changes, but it does from time to time.  So we could possibly cache this in ``redis``.  Perhaps invalidate it when a ``create`` is done on the table.  I'll see if that's straightforward...

File: ui/helpers/job.js (hunk 1 above)
Comment (camd) [reply ui/helpers/job.js]: Yeah, I'll remove JQuery usages in a follow-up PR>

**Comment by camd** [comment]
Tested by the Sheriff's on Prototype and they found no issues.

**Pull request merged**